### PR TITLE
Unit system

### PR DESCRIPTION
**Pull request by brryan**
## Background

We need to be able to set up a common conversion factor between code and physical units

## Description of Changes

- [x] Add `ArtemisUtils;:Units` class that contains unit conversions and physical constants in the physical unit system, set up from parameter input
- [x] Multiple ways to initialize code units (length/time/mass, star mass + orbital radius)
- [x] Ability to modify physical constants manually 
- [x] Report unit conversion in params so dump files include this information for analysis script usage?

## Checklist

- [x] New features are documented
- [ ] Tests added for bug fixes and new features
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp]: I'd slightly prefer the input block be `<units>` and not `<artemis/units>` since we don't really have `artemis/*` blocks.

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp:70]: I don't know if code units is `kb=1` or `Rgas=1`. I tend to use `Rgas=1` (and I think the default `cv` assumes that right now).

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp]: Should we be grabbing all the constants in parthenon's constants?

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp]: I would call these total mass and reference radius

File: src/utils/units.hpp
Comment (adamdempsey90) [inline src/utils/units.hpp:41]: Should these all be accessible on device?

File: src/nbody/nbody.cpp
Comment (adamdempsey90) [inline src/nbody/nbody.cpp]: This should really be GM not G

File: src/nbody/nbody.cpp
Comment (adamdempsey90) [inline src/nbody/nbody.cpp]: This should really be GM not G

File: src/nbody/nbody.cpp
Comment (adamdempsey90) [inline src/nbody/nbody.cpp:74]: Other packages like viscosity and maybe drag grab GM from here. So they'll need to be changed as well.

**Review by adamdempsey90 — COMMENTED** [review]
How are we going to access the units on device?

File: src/utils/units.cpp
Comment (pdmullen) [inline src/utils/units.cpp]: I think it would be really hard to handle all combinations... but it would be nice to also do something like specify a density unit and length unit.  Wonder if there is a clever way of doing something like that.

Comment (brryan) [reply src/utils/units.cpp]: I think you would just have to find a descriptive name for that approach for describing units and then compare `unit_specifier_` with that and then look for the appropriate input parameter constants. No real limit on how many such specifiers we can have, but I don't immediately see anything more clever than that.

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp]: My feeling was that `artemis/units` is the appropriate scope for these parameters, since 1) the units are used artemis-wide 2) other parthenon codes might have their own notion of `units` input parameters and this would avoid crosstalk between packages in different codes. 

We already have `artemis/` parameters so I don't see a style issue with having blocks as well there, but if you foresee issues with doing this lets keep discussing.

File: src/utils/units.cpp
Comment (pdmullen) [reply src/utils/units.cpp:70]: I'd personally prefer never to think about $R$ 😬  My brain is too hardwired to:

rho k T / (mu mp) == gm1 rho e

File: src/utils/units.cpp
Comment (brryan) [inline src/utils/units.cpp]:
```suggestion
  kb_code_ = kb_ * std::pow(time_, 2) / mass_ * std::pow(length_, -2); // 1 K = 1 code unit temp
```

**Comment by adamdempsey90** [comment]
Two design comments. 

(1) It almost feels like your "units" struct should really be called `PhysicalConstants` or something like that. The units part could maybe be separated into functions that do the conversions, e.g., `ConvertOpacityToCGS`. 

(2) I don't know if we need a full `<artemis/units>` node. Following the coordinate specification maybe just something in `<artemis>` directly,

```
<artemis>
units = cgs  # cgs, ppd, user
length = ...
mass   = ...
time    = ...
```
I also don't think we necessarily need the ability to specify every physical constant in the input deck. I would like to discourage that if at all possible.

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp:70]: Default units here is `kb = mp = 1`; I think `kb` makes more sense in this case because singularity-eos interacts with the specific heat capacity `<c_v> = erg / (g * K) -> c_v = k_b / (gam - 1) mu m_p`.

You are right that I need to actually calculate `c_v` correctly as part of this PR

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp]: I was just taking what we need; parthenon's constants is pretty thorough. Also because this class is runtime and be copied to device, it probably is best to keep it as small as possible.

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp]: Yes good point, in particular `total_mass` is actually correct based on how `nbody` already works. Fixed.

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp:41]: Yes; done. Class can only be constructed on device though because it interacts with `ParameterInput`. Device-usable copy constructor provided for getting this into kernels though.

**Comment by brryan** [comment]
> Two design comments.
> 
> (1) It almost feels like your "units" struct should really be called `PhysicalConstants` or something like that. The units part could maybe be separated into functions that do the conversions, e.g., `ConvertOpacityToCGS`.
> 
> (2) I don't know if we need a full `<artemis/units>` node. Following the coordinate specification maybe just something in `<artemis>` directly,
> 
> ```
> <artemis>
> units = cgs  # cgs, ppd, user
> length = ...
> mass   = ...
> time    = ...
> ```
> 
> I also don't think we necessarily need the ability to specify every physical constant in the input deck. I would like to discourage that if at all possible.

1) Yeah I think really this should be two different classes (now done). Only slight issue is a circular dependency with solar mass/AU in setting the unit system via reference mass/radius that I had to dance around. 

2) I'll defer to you on whether `units` is a node; it's gone now.

3) Yeah I removed the ability to customize physical constants. This was always intended as a "for experts only" option but it doesn't even work right now with our made up constants for e.g. the rad shocktubes because that's constexpr so I'm deferring the custom constants issue now.

File: src/nbody/nbody.cpp
Comment (brryan) [reply src/nbody/nbody.cpp]: The `M` here is the sum of the particles' `p.m` values right? So should this value be set in `NBodySetup` as `units.GetGCode() * mtot`?

Comment (adamdempsey90) [reply src/nbody/nbody.cpp]: Since `GM` is really set at startup and put into params, I don't think we need any units stuff in places where it's used. When it's initially defined it should probably just be `GM = units.GetGCode() * units.GetMCode() * GM`.

File: src/utils/units.hpp
Comment (adamdempsey90) [inline src/utils/units.hpp]: Other constants I'd like to see are:

- [ ]  `amu` and/or `mp` (cv = kb/(mu * amu *(g-1)))
- [ ]   `ev`
- [ ]  `year` 

At some point we'll also want  electron mass, but not at the moment. 

There's also `pc` if we want to be supportive AGN disks

File: src/gas/gas.cpp
Comment (adamdempsey90) [inline src/gas/gas.cpp]: This should be `mmw * amu`?

Also should we allow the user to specify `cv` separately from `mmw`, but default `cv` to this value?

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp]: Others that may be useful are:
- [x] opacity
- [x] energy density and/or pressure
- [x] velocity
- [x] whatever units `cv` is

File: src/utils/units.hpp
Comment (adamdempsey90) [inline src/utils/units.hpp:46]: What do you think about these functions taking in the quantity and returning the converted quantity, as opposed to returning conversion factors and the caller applying them.

File: src/utils/units.hpp
Comment (adamdempsey90) [inline src/utils/units.hpp]: A unit system for PPDs would be `Msolar` `AU` `yr/(2*pi)` and `K`. I think the CGS temperature unit makes it not "scalefree"? I would even call this the `ppd` unit system.

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp:19]: For Jupiter focused sims, would `Rjup` and `Mjup` also be useful to have?

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp]: Added `amu`, `ev`, `year`, `pc`, good suggestions :+1:

File: src/gas/gas.cpp
Comment (brryan) [reply src/gas/gas.cpp]: Yeah better to specify this in units of `amu` than `g`. 

We might want to *also* allow the user to specify non-ideal cvs, specifically the Marshak EOS, so maybe we do want a `model` parameter here, e.g. `model = mmw, cv, marshak`. I'll try that.

Comment (adamdempsey90) [reply src/gas/gas.cpp]: I'd suggest just doing, 

```
const Real mmw = pin->GetOrAddReal("gas", "mmw", 1.);
const Real cv = pin->GetOrAddReal("gas", "cv", constants.GetKBCode() / (mmw * constants.GetAMUCode() * (gamma - 1.)));
```

We could back out the `mmw` if the user specified just `cv`, but I think in cases like that `mmw` is probably meaningless. 

Also, can we call `mmw` either `mu` (which is what it's called in the astro literature) or `abar` (which is what it's called in the lab codes)?

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp:46]: I think this is a design choice that doesn't warrant a rewrite at this point. The one issue with the returning-converted-quantities approach is that you *need* functions for every kind of quantity you want to convert, unless you want to break the design if necessary by doing like
```c++
double l_unit = ConvertLengthCodeToCGS(1.);
double t_unit = ConvertTimeCodeToCGS(1.);
// Use some combination of powers of l_unit and t_unit to convert your custom quantity
```

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp]: I'm not sure we need unit systems beyond scalefree and cgs; these really only serve to have scalefree problems with trivial physical constants, or have a conversion to physical units. Whether we convert to length in cm or length in m or length in AU is already handled by the conversion between code and physical units,.

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp:19]: Agreed -- adding now

File: src/gas/gas.cpp
Comment (brryan) [reply src/gas/gas.cpp]: Switched `mmw` to `mu`, added optional `cv` override. I added logic so the user cant supply both `mu` and `cv` to avoid inconsistencies.

I was wrong about also supporting the Marshak EOS here, that will have to be a separate singularity-eos variant.

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp]: Done :+1:

File: src/utils/units.hpp
Comment (adamdempsey90) [reply src/utils/units.hpp:46]: I agree my choice comes with needing functions for every kind of quantity. I think that comes with safety, though, since there is one implementation of a unit conversion, and not multiple implementations. If a developer needs a conversion function that doesn't exist yet, they can add it.

File: src/utils/units.hpp
Comment (adamdempsey90) [reply src/utils/units.hpp]: I suppose it's just a convenience in the input file if `units = ppd` populates all the `length`, `time`, `mass`, etc. for me.

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp:46]: As discussed today let's keep the current function signatures.

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp]: I added a `ppd` option for `<artemis> unit_system = ` with your proposed unit conversions so this convenience should actually be available that way. :+1:

File: src/nbody/nbody.cpp
Comment (brryan) [inline src/nbody/nbody.cpp]: @adamdempsey90 OK I'm still thinking about this. I understand now how you provide nbody masses in terms of a portion of the total GM, but I think it's more consistent to specify total GM in terms of a total mass, and use the existing G in code units from our constants for consistency. I chose solar masses as the current unit for the total mass here; is that reasonable or would you prefer grams? Let me know if I'm still not understanding how this `GM` works but I think at this point this is consistent with the existing need for other packages to have a GM, and also consistent with our constants and units in this PR.

File: src/nbody/nbody.cpp
Comment (brryan) [reply src/nbody/nbody.cpp:74]: `gm` is again a `param`

File: src/nbody/nbody.cpp
Comment (adamdempsey90) [reply src/nbody/nbody.cpp]: I would think you would just multiply by `GetMCode`, not `GetMsolarCode`. `G` and `M` are set in such a way that in code units `GM=1` (for the `ppd` system).

Comment (adamdempsey90) [reply src/nbody/nbody.cpp]: Or maybe user inputs `GM` in code units and the mass unit, and then `G` is set from those two.

Comment (brryan) [reply src/nbody/nbody.cpp]: `G` is just `G`, so I strongly prefer respecting the value of that constant provided by the underlying scale (either `G=1` for `scalefree` or `G=6.67ewhatever` for `cgs` rather than letting the user override this in a maybe unexpected way (same way this PR no longer allows users to provide custom physical constants). With that requirement the only path forward I see is to have the user provide the total gravitating mass here, and let Artemis add the `G`. 

We can allow any units for `mass_tot` here, `Msolar` seemed like a reasonable choice since that's more likely to lead to round numbers in input files, but if you think there's a better default I don't have a strong opinion, except that I think we don't want the default to be in code units because code units aren't really for the user to interact directly with in general; we don't want people to have to juggle length/mass/time conversions when they're writing input failes. Also, for the `scalefree` backend `constants.GetMSolar() = 1` so in that case the user is just inputting trivially in code units (like current input decks). There isn't a `constants.GetMCode()` method because that would just return `1` (really it would be `units.GetMassCodeToCode()` which we dont provide because it's `=1`).

File: src/nbody/nbody.cpp
Comment (brryan) [reply src/nbody/nbody.cpp]: discussion moved below, resolving this

File: src/nbody/nbody.cpp
Comment (brryan) [reply src/nbody/nbody.cpp]: As discussed this afternoon in person, let's remove the `orbit` units with custom radius/gravitating mass, keep `ppd` units, and keep the `M` input parameter for specifiying the `GM` normalization since this mass is independent of the unit systems and must be specified somewhere in `nbody` gravity.

**Comment by brryan** [comment]
@adamdempsey90 @pdmullen IMO this is ready to go; could I please get approvals, or more queries if there are outstanding issues? Thanks

File: src/gas/gas.cpp
Comment (adamdempsey90) [inline src/gas/gas.cpp:160]: Just to be clear, even if artemis is using cgs units, this is still called `NonCGSUnits` right?

File: src/nbody/nbody.cpp
Comment (adamdempsey90) [inline src/nbody/nbody.cpp]: Why change this from `GM`? `GM` should already be in code units, so it was fine before right?

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp]: I don't understand the difference between `artemis/specifier` and `artemis/units/type`? I thought it would just be 

```
<artemis>
...
units = scalefree  # scalefree, cgs, ppd
length = ..
mass = ...
time = ...
```

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp:58]: Conversion factors for all the other compound quantities like `pressure`, `cv`, opacity, conductivity, viscosity?

File: src/artemis.cpp
Comment (adamdempsey90) [inline src/artemis.cpp:59]: Why pass `artemis` to the `Units` constructor? You're storing the whole thing in params, so why store other members separately as well?

File: src/utils/units.cpp
Comment (adamdempsey90) [reply src/utils/units.cpp:58]: I see below that you just compute those conversion on the fly. Not sure why some are computed on the fly and some are stored, but not a big deal

File: src/utils/units.hpp
Comment (adamdempsey90) [inline src/utils/units.hpp:132]: I don't know if every actually need `h`?

File: src/utils/units.cpp
Comment (adamdempsey90) [inline src/utils/units.cpp:94]: Sorry, one more constant that is being used is the radiation constant and optionally the steffan-boltzman constant.

File: src/nbody/nbody.cpp
Comment (brryan) [reply src/nbody/nbody.cpp]: Yes this was my leftover mistake, thanks for catching this, fixed.

File: src/gas/gas.cpp
Comment (brryan) [reply src/gas/gas.cpp:160]: Yes, this might be better called `CustomUnits` on the singularity side. I made the choice to always include the unit conversions to halve the number of singularity variants we're carrying around.

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp]: `[scalefree, cgs]` and `[base, ppd]` are separate ideas; the first fixes the physical units through setting the values of fundamental constants (either all `1` or cgs values) and `[base, ppd]` is different ways of setting the conversion between physical and code units, with `base` requiring the user to set length, mass, and time conversions, and `ppd` is a default set of those conversions for convenience with no user-settable parameters.

But I didnt completely remove the `units` node here so I will turn this into `artemis/physical_units` and `artemis/unit_conversion`

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp:58]: I tried to store the factors the need `std::pow` but calculate the others. It's just a choice I made.

File: src/artemis.cpp
Comment (brryan) [reply src/artemis.cpp:59]: It's crucial to store HDF5-outputtable conversion factors for use in analysis scripts, we could store those conversion factors separately here and not pass the artemis package but this felt like it kept `artemis.cpp` less messy.

File: src/utils/units.hpp
Comment (brryan) [reply src/utils/units.hpp:132]: It may show up with energy <-> frequency for photons somewhere

File: src/utils/units.cpp
Comment (brryan) [reply src/utils/units.cpp:94]: Good point, `ar` added

**Review by pdmullen — APPROVED** [review]
A huge effort. Thanks @brryan!

**Pull request merged**
